### ISSUE TITLE
How to call exported function?

### ISSUE THREAD

**Issue by MasterSoft24** [Dewera/Lunar#22]
Hi! How I could call exported function from program who inject dll's?

**Comment by Dewera** [comment]
Map the DLL with it's headers, parse the export directory to get the address of said function, then execute it using some shellcode in the context of the process.

Whilst the library has the functionality to do this internally (and does do this during the mapping process,) I didn't make it public to avoid bloating the public interface.

**Comment by MasterSoft24** [comment]
I does

**Comment by Dewera** [comment]
Thanks for bringing this to my attention - I was actually just trying to figure out the same thing as I'm going to push through an update for static tls data in the next few days and was having issues with shellcode execution.

Basically, it seems I introduced a bug in a commit somewhere down that makes the shellcode execution method I'm using unreliable (however for some odd reason still works fine in some cases.)

I will try and get a patch for this out before I do the release (once I figure out what the issue actually is.)

**Comment by Dewera** [comment]
I think I found the issue. You can try the following in your version for the time being

Change the [following](https://github.com/Dewera/Lunar/blob/master/Lunar/Extensions/ProcessExtensions.cs#L30)

```c#
var status = Ntdll.NtCreateThreadEx(out var threadHandle, AccessMask.SpecificRightsAll | AccessMask.StandardRightsAll, IntPtr.Zero, process.SafeHandle, address, IntPtr.Zero, ThreadCreationFlags.HideFromDebugger | ThreadCreationFlags.SkipThreadAttach, 0, 0, 0, IntPtr.Zero);
```

To

```c#
var status = Ntdll.NtCreateThreadEx(out var threadHandle, AccessMask.SpecificRightsAll | AccessMask.StandardRightsAll, IntPtr.Zero, process.SafeHandle, address, IntPtr.Zero, 0, 0, 0, 0, IntPtr.Zero);
```

I'm hoping to swap the execution method to something else soon, but that should work on your copy for the time being

**Comment by MasterSoft24** [comment]
You are right. This change has fix bug with target process crash. But my function still not working. During library mapping I've catch exception in LoadDependencies procedure. I've avoid it 

```
            System.Xml.Linq.XDocument manifest = null; // _peImage.ResourceDirectory.GetManifest();

            try
            {
                manifest = _peImage.ResourceDirectory.GetManifest();
            }
            catch { }

            var activationContext = new ActivationContext(manifest, _processContext.Process);
```
But I dont know how correct it is.

**Comment by Dewera** [comment]
Can you send me the DLL so I can check that out and do a fix? Discord is Quin#4576

**Comment by MasterSoft24** [comment]
Thank you again. Your fixes for https://github.com/Dewera/Lunar/issues/22#issuecomment-805391269 were really helpful. Problem solved
